magpie/api/webhooks.py

```diff
@@ -0,0 +1,89 @@
+import multiprocessing
+
+from pyramid.threadlocal import get_current_registry
+import requests
+import transaction
+
+from magpie.api.schemas import UserWebhookErrorStatus
+from magpie.constants import MAGPIE_INI_FILE_PATH
+from magpie.db import get_db_session_from_config_ini
+from magpie import models
+from magpie.utils import get_logger, get_settings
+
+# List of keys that should be found for a single webhook item in the config
+WEBHOOK_KEYS = {
+    "name",
+    "action",
+    "method",
+    "url",
+    "payload"
+}
+
+# List of possible actions associated with a webhook
+WEBHOOK_CREATE_USER_ACTION = "create_user"
+WEBHOOK_DELETE_USER_ACTION = "delete_user"
+WEBHOOK_ACTIONS = [
+    WEBHOOK_CREATE_USER_ACTION,
+    WEBHOOK_DELETE_USER_ACTION
+]
+
+# These are the parameters permitted to use the template form in the webhook payload.
+WEBHOOK_TEMPLATE_PARAMS = ["user_name", "tmp_url"]
+
+HTTP_METHODS = ["GET", "HEAD", "POST", "PUT", "PATCH", "DELETE"]
+
+LOGGER = get_logger(__name__)
+
+
+def process_webhook_requests(action, params, update_user_status_on_error=False):
+    """
+    Checks the config for any webhooks that correspond to the input action, and prepares corresponding requests.
+
+    :param action: tag identifying which webhooks to use in the config
+    :param params: dictionary containing the required parameters for the request, they will replace templates
+                    found in the payload
+    :param update_user_status_on_error: update the user status or not in case of a webhook error
+    """
+    # Check for webhook requests
+    webhooks = get_settings(get_current_registry())["webhooks"][action]
+    if len(webhooks) > 0:
+        # Execute all webhook requests
+        pool = multiprocessing.Pool(processes=len(webhooks))
+        args = [(webhook, params, update_user_status_on_error) for webhook in webhooks]
+        pool.starmap_async(send_webhook_request, args)
+
+
+def send_webhook_request(webhook_config, params, update_user_status_on_error=False):
+    """
+    Sends a single webhook request using the input config.
+
+    :param webhook_config: dictionary containing the config data of a single webhook
+    :param params: dictionary containing the required parameters for the request, they will replace templates
+                    found in the payload
+    :param update_user_status_on_error: update the user status or not in case of a webhook error
+    """
+    # Replace each instance of template parameters if a corresponding value was defined in input
+    for template_param in WEBHOOK_TEMPLATE_PARAMS:
+        if template_param in params:
+            for k, v in webhook_config["payload"].items():
+                webhook_config["payload"][k] = v.replace("{" + template_param + "}", params[template_param])
```

Comment:
I would like better flexibility for the payload than a simple dict. For me the payload could be any valid json.
Ex. 1:
```
POST http://myfancyservice.com/users : [<user_name>]
```
Ex. 2:
```
POST http://myfancyservice.com/channels : {
    'announce' : [
        {
            'channel' : 'x',
            'message' : 'user <user_name> has just been added'
        },
    ] 
}
```
So you would need to : 

1. check for dict vs list vs other instance 
2. replace within values and keys 
3. recurse into dict and list value

Reply:
I will look into this!

Reply:
@dbyrns 
Just to be sure, with the json format, the first level of data has to be a dict. Then its values can have more flexibility, being either a list, a dict, or basic types (string, int, bool, ..). So I think your first example here would not work as it's just a list?

Reply:
I choose this example specifically because a top level array is in the json spec : https://stackoverflow.com/questions/3833299/can-an-array-be-top-level-json-text#:~:text=Yes%2C%20an%20array%20is%20legal,top%2Dlevel%20JSON%2Dtext.

Reply:
I think I understand my error. I needed to use the 'json' parameter instead of 'data', in the request method. It was causing problems with lists!

Reply:
@dbyrns 
Latest commits should fix the changes you requested!